Function specializer

hrkrshnn opened the pull request:
TODO:

- [x] Fix the issues about generating the same name.
- [x] Check why gas costs of abiv2_optimized is higher.
- [x] More tests.

hrkrshnn committed:
Implemented FunctionSpecializer

Optimiser step that specializes the function with its literal arguments.

hrkrshnn committed:
NameCollecter can now only collect VariableDeclarations

This is done to help with the optimizer step FunctionSpecializer

hrkrshnn committed:
Added FunctionSpecializer to the optimization suite

hrkrshnn committed:
Tests for FunctionSpecializer

hrkrshnn commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+	if (boost::algorithm::any_of(arguments, [](auto& _a) { return _a.has_value(); }))
+	{
+		auto newName = m_nameDispenser.newName(_f.functionName.name);
+		yulAssert(newName != _f.functionName.name, "Name dispenser returned the same name.");
```
Tests are failing because the name generated is the same.

axic commented on test/libyul/yulOptimizerTests/fullSuite/fibonacci.yul:
```diff
@@ -0,0 +1,39 @@
+//         sstore(7, 13)
+//         sstore(8, 21)
+//         sstore(9, 34)
+//         sstore(10, 55)
```
Nice!

axic replied:
Can you add a large `fib()` case to see if we have some resource exhaustion limit or it just becomes really slow to optimize?

chriseth replied:
Impressive!

hrkrshnn replied:
> Can you add a large fib() case to see if we have some resource exhaustion limit or it just becomes really slow to optimize?

Added a test for that `test/libyul/yulOptimizerTests/fullSuite/ackermann_function.yul`

chriseth commented on test/libyul/yulOptimizerTests/fullSuite/aztec.yul:
```diff
@@ -320,40 +320,41 @@
 //         return(0, 0x20)
 //     }
-//     function validatePairing(t2)
+//     function validatePairing_1113()
```
This is also nice!

chriseth commented on test/libsolidity/gasTests/abiv2_optimised.sol:
```diff
@@ -17,9 +17,9 @@ contract C {
-//   codeDepositCost: 578200
-//   executionCost: 613
-//   totalCost: 578813
+//   codeDepositCost: 624000
```
Deposit and execution costs both go up :(

chriseth commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,103 @@
+ *     sstore(a_1, b_1)
+ *   }
+ *
+ * Further optimization steps will be able to simplify the function even further. The optimization
```
repeated `further`

chriseth commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,103 @@
+{
+	/// A vector of function-call arguments. An element 'has value' if it's a literal, and the
+	/// corresponding Expression would be the literal.
+	using Arguments = std::vector<std::optional<Expression>>;
```
```suggestion
	using LiteralArguments = std::vector<std::optional<Expression>>;
```
?

chriseth replied:
Acutally wouldn't it be better to use a pointer to Expression?

chriseth replied:
Disregard the comment abount pointer.

hrkrshnn replied:
I changed the name to LiteralArguments

chriseth commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,103 @@
+	/// Given a function definition `_f` and its arguments `_arguments`, of which, at least one is a
+	/// literal, this function returns a new function with the literal arguments specialized.
+	///
+	/// Note that the return function definition will have new (and unique) names, for both the
```
```suggestion
	/// Note that the returned function definition will have new (and unique) names, for both the
```

chriseth commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,103 @@
+	/// literal, this function returns a new function with the literal arguments specialized.
+	///
+	/// Note that the return function definition will have new (and unique) names, for both the
+	/// function and variable declarations.
```
```suggestion
	/// function and variable declarations to retain the properties enforced by the Disambiguator.
```

chriseth commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,103 @@
+	/// function and variable declarations.
+	///
+	/// For example, if `_f` is the function `function f(a, b, c) -> d { sstore(a, b) }` and
+	/// `_arguments` is the vector of literals `{1, 2}`. The returned function could be:
```
```suggestion
	/// `_arguments` is the vector of literals `{1, 2, nullopt}`. The returned function could be:
```

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+
+/// Helper class that creates a copy of the function definition, replacing the names of the variable
+/// declaration with a new name.
+class FunctionCopier: public ASTCopier
```
Does it make sense to move that to the main AST Copier?

hrkrshnn replied:
Moved it to ASTCopier

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+	{
+		auto newName = m_nameDispenser.newName(_f.functionName.name);
+		yulAssert(newName != _f.functionName.name, "Name dispenser returned the same name.");
+		m_oldToNewMap[_f.functionName.name].emplace_back(make_pair(newName, arguments));
```
Does it make sense to search if the specialization is already present? You can use SyntacticallyEqual to compare.

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+		m_oldToNewMap[_f.functionName.name].emplace_back(make_pair(newName, arguments));
+
+		_f.functionName.name = newName;
+		_f.arguments = unusedFunctionsCommon::filter(
```
Isn't this just `boost::erase_remove(_f.arguments, nullopt)`?

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+		map<YulString, YulString>{}
+	);
+
+	FunctionDefinition newFunction = get<FunctionDefinition>(FunctionCopier{translatedNames}(_f));
```
Isn't there a utility already that creates a copy with new variable names?

hrkrshnn replied:
There is BodyCopier. But it is not supposed to handle `FunctionDefinition`. I don't think there is anything else that can work.

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+
+	FunctionDefinition newFunction = get<FunctionDefinition>(FunctionCopier{translatedNames}(_f));
+
+	// Function parameters that will be specialized inside the body is converted into a variable
```
```suggestion
	// Function parameters that will be specialized inside the body are converted into variable
```

chriseth commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,183 @@
+	FunctionDefinition newFunction = get<FunctionDefinition>(FunctionCopier{translatedNames}(_f));
+
+	// Function parameters that will be specialized inside the body is converted into a variable
+	// declaration.
```
```suggestion
	// declarations.
```

hrkrshnn committed:
Added a helper class FunctionCopier in ASTCopier

Helper class that creates a copy of the function definition, replacing the names of the variable
declaration with a new name.

chriseth commented on test/libyul/yulOptimizerTests/fullSuite/ackermann_function.yul:
```diff
@@ -0,0 +1,559 @@
+//     function A_294_20862() -> ret
+//     {
+//         switch 1
+//         case 0 { ret := add(ret, 1) }
```
Interesting - why doesn't it use `ret = 0`?

hrkrshnn replied:
I think we reached the `maxRounds` for `runSequenceUntilStable`.

chriseth commented on test/libyul/yulOptimizerTests/fullSuite/ackermann_function.yul:
```diff
@@ -0,0 +1,559 @@
+//         switch 1
+//         case 0 { ret := add(ret, 1) }
+//         default {
+//             let _1 := add(1, not(0))
```
Why is this not simplified to `0`?

leonardoalt commented on libyul/optimiser/ASTCopier.h:
```diff
@@ -117,5 +118,22 @@ std::vector<T> ASTCopier::translateVector(std::vector<T> const& _values)
 }
 
+/// Helper class that creates a copy of the function definition, replacing the names of the variable
+/// declaration with a new name.
```
```suggestion
/// declarations with new names.
```

leonardoalt commented on libyul/optimiser/ASTCopier.h:
```diff
@@ -117,5 +118,22 @@ std::vector<T> ASTCopier::translateVector(std::vector<T> const& _values)
+	using ASTCopier::operator();
+	YulString translateIdentifier(YulString _name) override;
+private:
+	/// A mapping between old and new names. We replace the names of variable declaration contained
```
```suggestion
	/// A mapping between old and new names. We replace the names of variable declarations contained
```

leonardoalt commented on libyul/optimiser/ASTCopier.h:
```diff
@@ -117,5 +118,22 @@ std::vector<T> ASTCopier::translateVector(std::vector<T> const& _values)
+	YulString translateIdentifier(YulString _name) override;
+private:
+	/// A mapping between old and new names. We replace the names of variable declaration contained
+	/// in the mapping with its new name.
```
```suggestion
	/// in the mapping with their new names.
```

leonardoalt commented on test/libyul/yulOptimizerTests/functionSpecializer/simple.yul:
```diff
@@ -0,0 +1,28 @@
+//
+// {
+//     f_1()
+//     function f_1()
```
What is the benefit in this case?

hrkrshnn replied:
The benefit is only theoretical, that some other step might be able to use it.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,149 @@
+#include <libyul/YulString.h>
+#include <libsolutil/CommonData.h>
+
+#include <boost/algorithm/cxx11/any_of.hpp>
```
Can this be replaced by any of the new range stuff?

hrkrshnn replied:
We can replace `boost::range::any_of` with `ranges::any_of`, but it feels weird using the same function from range-v3 instead of boost.

hrkrshnn replied:
But I'll replace it here.

leonardoalt replied:
I think the advantage is that once we have c++20 it's kind of "done". And this way we can keep reducing our boost footprint :p

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,149 @@
+
+	if (boost::algorithm::any_of(arguments, [](auto& _a) { return _a.has_value(); }))
+	{
+		auto const& oldName = _f.functionName.name;
```
Could `move` here, since you're reassigning to `_f.functionName.name` later

hrkrshnn replied:
Done.

chriseth commented on test/cmdlineTests/name_simplifier/output:
```diff
@@ -119,14 +125,20 @@ object "C_56" {
-            function storage_array_index_access$_t_struct$_S_storage(array, index) -> slot, offset
+            function storage_array_index_access$_t_struct$_S_storage_1219() -> slot, offset
+            {
+                if iszero(lt(slot, 0x02)) { invalid() }
```
It's weird that this is not optimized.

hrkrshnn replied:
https://gist.github.com/hrkrshnn/71c49684c5ae60974f5bed90f40fb73a is the AST after it's running the "[]" part of full suite. Using yulopti, we don't get to the point where the condition is getting simplified to 0.

hrkrshnn replied:
Also has nothing to do with reaching the max for `maxRounds`. In fact 'runSequenceUntilStable' stops after just 4 rounds.

hrkrshnn commented on test/libyul/yulOptimizerTests/fullSuite/reserved_identifiers.yul:
```diff
@@ -16,14 +16,160 @@
 //     {
-//         let dataoffset_ := datasize_(7)
-//         sstore(dataoffset_, g(9))
+//         let linkersymbol_1 := 0
```
This test case has become much worse.

chriseth replied:
Can you detect recursion?

hrkrshnn replied:
We could get that info from `SideEffectsCollector`. But we shouldn't just skip recursive functions, I think. If we do, then our fibonacci optimization will not work any more.

Ideally, the way to go ahead is by backtracking.

ekpyron replied:
Maybe it's still an option to just disable it on recursive functions for now, even though it hurts ;-), and try to work towards getting it back in a second step?

chriseth replied:
I would agree here. Recursive functions should be extremely rare.

bshastry commented on libsolutil/CommonData.h:
```diff
@@ -93,10 +93,8 @@ template <class T>
-	else
-		ret += std::move(_b);
+	assert(&_a != &_b);
+	ret += std::move(_b);
```
This means that the + operator may no longer be used on self, right? like `v += v;` where `v` is of type `vector<T>`?

hrkrshnn replied:
Not sure about that. Something like `v += move(v)` cannot be used. But does that even make sense?

bshastry replied:
Ah right. Wonder why the if case existed before. Caught my attention because this is in utilities which may be used elsewhere?

ekpyron replied:
It's ``+`` not ``+=``... so it'd be ``move(v) + move(v)``... which is not too likely itself :-)...
But then again if a template function that uses perfect forwarding does this on some arguments and that template is called with an rvalue... I wouldn't say this *can't* happen :-).. I stumbled across this myself before, but then didn't touch it here and ended up just not using ``+`` in that case...
What triggered this here?
```

	newFunction.body.statements =
		move(missingVariableDeclarations) + move(newFunction.body.statements);
```
probably?
```
	newFunction.body.statements = move(missingVariableDeclarations);
    newFunction.body.statements += move(newFunction.body.statements);
```
would probably work without the change here...

bshastry commented:
Is this PR review ready?

hrkrshnn commented:
@bshastry You can review it. There are two open problems now.

1. https://github.com/ethereum/solidity/pull/10418#discussion_r560091140
2. There are cases when the optimization rule can make the code size much larger, for example, recursive function that really can't be simplified. How do we detect them?

bshastry commented on test/libyul/YulOptimizerTestCommon.cpp:
```diff
@@ -173,6 +174,13 @@ bool YulOptimizerTestCommon::runStep()
+		FunctionHoister::run(*m_context, *m_object->code);
+
+		FunctionSpecializer::run(*m_context, *m_object->code);
+	}
```
Needs tiny rebase due to https://github.com/ethereum/solidity/pull/10811

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/ASTWalker.h>
```
Could you please order includes alphabetically?

hrkrshnn replied:
Fixed.

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+
+#include <optional>
+#include <vector>
+#include <map>
```
Non alphabetically sorted

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+{
+
+/**
+ * FunctionSpecializer: Optimiser step that specializes the function with its literal arguments.
```
UK spelling i.e., `FunctionSpecialiser`? Not sure if we've done this uniformly but `optimiser` seems to be that way.

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+ *   function f_1(a_1) {
+ *     let b_1 := 5
+ *     sstore(a_1, b_1)
+ *   }
```
Let's say we have the following sequence of calls

```
f(x, y)
f(x, 5)
f(5, y)
f(5, 5)
```

Does this generate four copies of the same function (not exactly copies but specialisations). If yes, doesn't this increase deployment cost? Of course, deployment gas cost is a one time thing but just wanted to bring this up if it has not already been considered.

hrkrshnn replied:
It does increase the deployment cost, provided no further optimizations could be made. Hopefully, specializing would improve further optimizations. But a proper way to deal with it is to have backtracking, i.e., some form of: let the sequence execute, and we check in the end if the costs were decreased. If not, skip applying the sequence. What do you think should be a fix?

bshastry replied:
Sounds good. I think it is fine if deployment cost increases but run time cost reduces. I think the latter is a priority especially for code that is not frequently updated.

One solution is what you suggest. The other could be a heuristic threshold for specialisation like `maxSpecialisedFunctions` per function or even per source unit?

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+	static constexpr char const* name{"FunctionSpecializer"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	using ASTModifier::operator();
```
Is the using declaration needed?

hrkrshnn replied:
Yes. Without it, you would get errors like these:

```
error: no match for call to ‘(solidity::yul::FunctionSpecializer) (solidity::yul::Block&)’
  126 |  f(_ast);
```

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+		m_dialect(_dialect)
+	{}
+	/// Returns a vector of Expressions, where the index `i` is an expression if the function's
+	/// `i`-th argument can be specialized.
```
```suggestion
	/// `i`-th argument can be specialized, nullopt otherwise.
```

hrkrshnn replied:
Done

bshastry commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,104 @@
+	///   }
+	FunctionDefinition specialize(
+		FunctionDefinition const& _f,
+		YulString _newName,
```
Is `_newName` the new name assigned to the specialised function i.e., `f_1` in the example?

bshastry replied:
If so, perhaps add this to the documentation?

hrkrshnn replied:
Added this.

bshastry commented on test/libsolidity/gasTests/abiv2_optimised.sol:
```diff
@@ -17,9 +17,9 @@ contract C {
-//   totalCost: 589419
+//   codeDepositCost: 642000
+//   executionCost: 676
+//   totalCost: 642676
```
One would expect this is due to specialised functions but none of the functions in the test contract are actually called,

hrkrshnn replied:
Some of the abi functions are getting specialized here. I can show a diff later.

bshastry reviewed (commented):
First set of comments. Will take a closer look at tests next.

axic commented on test/libyul/yulOptimizerTests/fullSuite/ackermann_function.yul:
```diff
@@ -0,0 +1,559 @@
+    // Too many unrolling needed. In arbitrary precision, the value is 2**65536 - 3.
+    sstore(2, A(4, 2))
+
+    function A(m, n) -> ret {
```
Can you create a second test with implementing the ackermann function using if+leave?

Like:
```
function A(m, n) -> ret {
  if eq(m, 0) {
    ret := add(n, 1)
    leave
  }

  if eq(n, 0) {
    ret := A(sub(m, 1), 1)
    leave
  }

  ret := A(sub(m, 1), A(m, sub(n, 1)))
}
```

hrkrshnn replied:
Added the test `ackermann_function_if.yul`

leonardoalt commented:
Needs rebase. @hrkrshnn were the problems discussed after the first round resolved?

hrkrshnn commented:
@leonardoalt I want to wait for the "isoltest gas" PR to be merged first (https://github.com/ethereum/solidity/pull/10863) Then the review can be more convincing.

Also, there is an open problem on whether we should specialize recursive functions or not. Currently, we do.

hrkrshnn committed:
Isoltest gas update for FunctionSpecializer

hrkrshnn commented:
<details><summary>Click for a table of gas differences</summary>

| file name | IR-Opti | Legacy-Opti | Legacy|
|-----------|---------|-------------|-------|
| interface_inheritance_conversions.sol |      0.669% |      0.000% |      0.000% |
| abiEncoderV1/abi_decode_v2_storage.sol |     -0.246% |      0.000% |      0.000% |
| abiEncoderV1/abi_encode_calldata_slice.sol |     -0.942% |      0.000% |      0.000% |
| abiEncoderV1/struct/struct_storage_ptr.sol |     -3.991% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2.sol |     -0.352% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2_in_modifier_used_in_v1_contract.sol |      0.000% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_calldata_slice.sol |     -0.942% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol |      2.118% |      0.000% |      0.000% |
| abiEncoderV2/calldata_array.sol |     -0.161% |      0.000% |      0.000% |
| abiEncoderV2/storage_array_encoding.sol |     -0.125% |      0.000% |      0.000% |
| abiencodedecode/abi_decode_simple_storage.sol |     -0.074% |      0.000% |      0.000% |
| array/fixed_array_cleanup.sol |     -0.011% |      0.000% |      0.000% |
| array/reusing_memory.sol |     -0.148% |      0.000% |      0.000% |
| array/arrays_complex_from_and_to_storage.sol |     -0.184% |      0.000% |      0.000% |
| array/byte_array_storage_layout.sol |     -0.170% |      0.000% |      0.000% |
| array/byte_array_transitional_2.sol |     -0.261% |      0.000% |      0.000% |
| array/bytes_length_member.sol |     -0.052% |      0.000% |      0.000% |
| array/create_memory_array.sol |    -11.425% |      0.000% |      0.000% |
| array/dynamic_array_cleanup.sol |     -0.040% |      0.000% |      0.000% |
| array/dynamic_arrays_in_storage.sol |     -0.161% |      0.000% |      0.000% |
| array/dynamic_multi_array_cleanup.sol |     -6.073% |      0.000% |      0.000% |
| array/fixed_arrays_as_return_type.sol |    -15.102% |      0.000% |      0.000% |
| array/function_array_cross_calls.sol |     -1.522% |      0.000% |      0.000% |
| array/copying/array_copy_calldata_storage.sol |     -2.640% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint128.sol |     -3.714% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint40.sol |     -7.728% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage.sol |     -7.241% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage_packed.sol |     -3.746% |      0.000% |      0.000% |
| array/copying/array_copy_different_packing.sol |     -0.244% |      0.000% |      0.000% |
| array/copying/array_copy_including_array.sol |     -0.134% |      0.000% |      0.000% |
| array/copying/array_copy_nested_array.sol |     -0.142% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_different_base.sol |     -0.168% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_different_base_nested.sol |     -2.170% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol |     -3.906% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_dynamic.sol |     -0.998% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_static.sol |     -2.508% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_struct.sol |     -8.317% |      0.000% |      0.000% |
| array/copying/array_copy_storage_to_memory_nested.sol |     -1.127% |      0.000% |      0.000% |
| array/copying/array_copy_target_leftover.sol |     -1.891% |      0.000% |      0.000% |
| array/copying/array_copy_target_leftover2.sol |     -1.265% |      0.000% |      0.000% |
| array/copying/array_copy_target_simple.sol |     -0.418% |      0.000% |      0.000% |
| array/copying/array_copy_target_simple_2.sol |     -2.089% |      0.000% |      0.000% |
| array/copying/array_nested_calldata_to_storage.sol |     -0.803% |      0.000% |      0.000% |
| array/copying/array_nested_memory_to_storage.sol |     -0.804% |      0.000% |      0.000% |
| array/copying/array_of_struct_calldata_to_storage.sol |     -0.179% |      0.000% |      0.000% |
| array/copying/array_of_struct_memory_to_storage.sol |     -0.475% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol |     -0.134% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol |     -1.079% |      0.000% |      0.000% |
| array/copying/array_storage_multi_items_per_slot.sol |     -1.080% |      0.000% |      0.000% |
| array/copying/arrays_from_and_to_storage.sol |     -0.082% |      0.000% |      0.000% |
| array/copying/bytes_inside_mappings.sol |     -0.036% |      0.000% |      0.000% |
| array/copying/bytes_storage_to_storage.sol |     -0.288% |      0.000% |      0.000% |
| array/copying/calldata_array_dynamic_to_storage.sol |     -2.563% |      0.000% |      0.000% |
| array/copying/copy_byte_array_in_struct_to_storage.sol |     -0.234% |      0.000% |      0.000% |
| array/copying/copy_byte_array_to_storage.sol |     -0.771% |      0.000% |      0.000% |
| array/copying/copy_function_storage_array.sol |     -0.312% |      0.000% |      0.000% |
| array/copying/copy_removes_bytes_data.sol |     -0.040% |      0.000% |      0.000% |
| array/copying/memory_dyn_2d_bytes_to_storage.sol |     -0.523% |      0.000% |      0.000% |
| array/copying/storage_memory_nested.sol |     -2.472% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_bytes.sol |     -0.158% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_from_pointer.sol |     -2.472% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_struct.sol |     -1.105% |      0.000% |      0.000% |
| array/copying/storage_memory_packed_dyn.sol |    -12.661% |      0.000% |      0.000% |
| array/delete/bytes_delete_element.sol |     -0.297% |      0.000% |      0.000% |
| array/delete/delete_storage_array_packed.sol |     -0.260% |      0.000% |      0.000% |
| array/pop/array_pop_array_transition.sol |     -3.700% |      0.000% |      0.000% |
| array/pop/array_pop_uint16_transition.sol |     -1.279% |      0.000% |      0.000% |
| array/pop/array_pop_uint24_transition.sol |     -1.158% |      0.000% |      0.000% |
| array/pop/byte_array_pop_copy_long.sol |     -0.144% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty.sol |     -0.938% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty_garbage_ref.sol |     -0.404% |      0.000% |      0.000% |
| array/pop/byte_array_pop_masking_long.sol |     -0.162% |      0.000% |      0.000% |
| array/push/array_push.sol |     -4.593% |      0.000% |      0.000% |
| array/push/array_push_nested_from_calldata.sol |     -0.152% |      0.000% |      0.000% |
| array/push/array_push_packed_array.sol |     -0.550% |      0.000% |      0.000% |
| array/push/array_push_struct.sol |     -0.643% |      0.000% |      0.000% |
| array/push/array_push_struct_from_calldata.sol |     -0.258% |      0.000% |      0.000% |
| array/push/byte_array_push_transition.sol |     -0.447% |      0.000% |      0.000% |
| array/push/push_no_args_2d.sol |     -0.027% |      0.000% |      0.000% |
| array/push/push_no_args_bytes.sol |     -1.061% |      0.000% |      0.000% |
| constructor/no_callvalue_check.sol |     -0.046% |      0.000% |      0.000% |
| constructor/arrays_in_constructors.sol |      1.389% |      0.000% |      0.000% |
| constructor/bytes_in_constructors_packer.sol |      1.974% |      0.000% |      0.000% |
| functionCall/failed_create.sol |      4.780% |      0.000% |      0.000% |
| functionCall/mapping_array_internal_argument.sol |     -0.829% |      0.000% |      0.000% |
| functionTypes/store_function.sol |     -3.084% |      0.000% |      0.000% |
| immutable/multi_creation.sol |     -0.062% |      0.000% |      0.000% |
| salted_create/salted_create.sol |     -0.000% |      0.000% |      0.000% |
| salted_create/salted_create_with_value.sol |     -3.267% |      0.000% |      0.000% |
| storage/packed_storage_structs_bytes.sol |     -0.230% |      0.000% |      0.000% |
| structs/struct_copy.sol |     -0.051% |      0.000% |      0.000% |
| structs/struct_copy_via_local.sol |     -0.079% |      0.000% |      0.000% |
| structs/struct_memory_to_storage_function_ptr.sol |     -0.128% |      0.000% |      0.000% |
| structs/structs.sol |     -0.119% |      0.000% |      0.000% |
| structs/memory_structs_nested_load.sol |     -0.646% |      0.000% |      0.000% |
| structs/struct_containing_bytes_copy_and_delete.sol |     -0.063% |      0.000% |      0.000% |
| structs/struct_delete_storage_nested_small.sol |     -0.382% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_array.sol |     -4.234% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_arrays_small.sol |     -0.613% |      0.000% |      0.000% |
| structs/calldata/calldata_struct_with_nested_array_to_storage.sol |     -2.165% |      0.000% |      0.000% |
| structs/conversion/recursive_storage_memory.sol |     -1.466% |      0.000% |      0.000% |
| various/swap_in_storage_overwrite.sol |     -0.047% |      0.000% |      0.000% |
| various/destructuring_assignment.sol |     -2.792% |      0.000% |      0.000% |
| various/skip_dynamic_types_for_structs.sol |     -0.256% |      0.000% |      0.000% |
| viaYul/array_storage_length_access.sol |     -0.001% |      0.000% |      0.000% |
| viaYul/array_memory_index_access.sol |     -0.048% |      0.000% |      0.000% |
| viaYul/array_storage_index_access.sol |     -5.251% |      0.000% |      0.000% |
| viaYul/array_storage_index_boundary_test.sol |     -0.294% |      0.000% |      0.000% |
| viaYul/array_storage_index_zeroed_test.sol |     -4.871% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty.sol |     -0.349% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty_length_address.sol |     -0.299% |      0.000% |      0.000% |
| viaYul/array_storage_push_pop.sol |     -0.231% |      0.000% |      0.000% |
| externalContracts/deposit_contract.sol |      0.145% |      0.000% |      0.000% |
| externalContracts/snark.sol |    -10.877% |      0.000% |      0.000% |
| inheritance/address_overload_resolution.sol |    -11.323% |      0.000% |      0.000% |
| inheritance/inherited_function_calldata_calldata_interface.sol |     -4.078% |      0.000% |      0.000% |
| inheritance/inherited_function_calldata_memory_interface.sol |      5.545% |      0.000% |      0.000% |

</details>

hrkrshnn commented:
This should be ready now. Posted a table of gas differences, and it looks mostly nice. The problem is that there can be pathological cases like the `ackermann.yul` where this step would produce horrible IR. But detecting them would require backtracking.

I would propose to add them to the full suite and perhaps add a command line switch to disable this step. What do you think @chriseth @axic

chriseth commented on libsolidity/interface/OptimiserSettings.h:
```diff
@@ -45,7 +45,7 @@ struct OptimiserSettings
 			"xarrcL"                   // Turn into SSA again and simplify
 			"gvif"                     // Run full inliner
-			"CTUcarrLsTOtfDncarrIulc"  // SSA plus simplify
+			"CTUcarrLsTFOtfDncarrIulc" // SSA plus simplify
```
Can you explain why you put the F in exactly that place?

hrkrshnn replied:
Mainly I wanted to have it after the full inliner. And putting after Literal Rematerialiser (T) would make sense.

hrkrshnn committed:
Changelog and adding the name to the doc

hrkrshnn committed:
FunctionSpecializer: skip specializing recursive functions

This avoids potential pathological behaviour, like in Ackermann function.

stackenbotten commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,159 @@
+{
+	FunctionSpecializer f{
+		CallGraphGenerator::callGraph(_ast).recursiveFunctions(),
+		 _context.dispenser,
```
Coding style error

stackenbotten commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,159 @@
+	FunctionSpecializer f{
+		CallGraphGenerator::callGraph(_ast).recursiveFunctions(),
+		 _context.dispenser,
+		 _context.dialect
```
Coding style error

hrkrshnn commented:
The last commit adds code to avoid specializing recursive functions. So the Fibonacci and Ackermann example would be untouched.

leonardoalt commented on libyul/optimiser/ASTCopier.h:
```diff
@@ -117,5 +118,22 @@ std::vector<T> ASTCopier::translateVector(std::vector<T> const& _values)
+	FunctionCopier(
+		std::map<YulString, YulString> _translations
+	):
+		m_translations(_translations)
```
```suggestion
		m_translations(std::move(_translations))
```

leonardoalt commented on libyul/optimiser/ASTCopier.h:
```diff
@@ -117,5 +118,22 @@ std::vector<T> ASTCopier::translateVector(std::vector<T> const& _values)
+private:
+	/// A mapping between old and new names. We replace the names of variable declarations contained
+	/// in the mapping with their new names.
+	std::map<YulString, YulString> const m_translations;
```
If this is `const` and relies on the map given in the constructor, why not make it `const&`?

leonardoalt commented on test/cmdlineTests/evm_to_wasm_break/output:
```diff
@@ -24,68 +24,64 @@ object "object" {
         {
-            let _1 := 0
-            let x, x_1, x_2, x_3 := calldataload(_1, _1, _1, _1)
+            let x, x_1, x_2, x_3 := calldataload()
```
I would expect `calldataload` to be sort of a dialect built-in. Are those also potentially modified?

hrkrshnn replied:
`calldataload` is not a built in for wasm.

leonardoalt replied:
Ah ok

leonardoalt commented on test/libyul/YulOptimizerTestCommon.cpp:
```diff
@@ -16,6 +16,7 @@
 */
 // SPDX-License-Identifier: GPL-3.0
 
+#include <boost/system/error_code.hpp>
```
Why this include?

hrkrshnn replied:
Not sure :) Probably automatically added and I missed this.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,113 @@
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	using ASTModifier::operator();
+	virtual void operator()(FunctionCall& _f) override;
```
```suggestion
	void operator()(FunctionCall& _f) override;
```

leonardoalt commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,113 @@
+	/// Note that the returned function definition will have new (and unique) names, for both the
+	/// function and variable declarations to retain the properties enforced by the Disambiguator.
+	///
+	/// For example, if `_f` is the function `function f(a, b, c) -> d { sstore(a, b) }` and
```
```suggestion
	/// For example, if `_f` is the function `function f(a, b, c) { sstore(a, b) }` and
```
Or return something? Otherwise `d` is unused

hrkrshnn replied:
I think it's better to leave this. I wanted the example to show that the return parameter also have to be renamed, without having to make the example longer.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,113 @@
+	///
+	/// For example, if `_f` is the function `function f(a, b, c) -> d { sstore(a, b) }` and
+	/// `_arguments` is the vector of literals `{1, 2, nullopt}` and the parameter, `_newName` has
+	/// value `f_1`. Then, the returned function could be:
```
```suggestion
	/// For example, if `_f` is the function `function f(a, b, c) -> d { sstore(a, b) }`,
	/// `_arguments` is the vector of literals `{1, 2, nullopt}` and @param `_newName` has
	/// value `f_1`, the returned function could be:
```

leonardoalt commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,113 @@
+		FunctionSpecializer::LiteralArguments _arguments
+	);
+
+	/// A mapping between the old function-name and a pair of new function-name and its arguments.
```
```suggestion
	/// A mapping between the old function-name and a pair of new function name and its arguments.
```

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,157 @@
+{
+	ASTModifier::operator()(_f);
+
+	// When backtracking is implemented, the restriction of recursive functions can be lifted.
```
```suggestion
	// TODO When backtracking is implemented, the restriction of recursive functions can be lifted.
```
Just so it's easier to find later on

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,157 @@
+	map<YulString, YulString> translatedNames = applyMap(
+		NameCollector{_f, NameCollector::OnlyVariables}.names(),
+		[&](auto& _name) -> pair<YulString, YulString>
+		{
```
I usually put `{` in the previous line in lambdas like this, but fine with either.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,157 @@
+		{
+			return make_pair(_name, m_nameDispenser.newName(_name));
+		},
+		map<YulString, YulString>{}
```
Do you have to specify it?

hrkrshnn replied:
Yes. The default output container is a vector.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,157 @@
+	// Function parameters that will be specialized inside the body are converted into variable
+	// declarations.
+	vector<Statement> missingVariableDeclarations;
+	for (size_t i = 0; i < _f.parameters.size(); ++i)
```
Could be replaced by ranges enumerate

hrkrshnn replied:
I replaced them by ranges enumerate.

leonardoalt commented on libyul/optimiser/FunctionSpecializer.h:
```diff
@@ -0,0 +1,113 @@
+	/// Note that at least one of the argument will have a literal value.
+	std::map<YulString, std::vector<std::pair<YulString, LiteralArguments>>> m_oldToNewMap;
+	/// We skip specializing recursive functions. Need backtracking to properly deal with them.
+	std::set<YulString> m_recursiveFunctions;
```
```suggestion
	std::set<YulString> const m_recursiveFunctions;
```

leonardoalt commented on libyul/optimiser/FunctionSpecializer.cpp:
```diff
@@ -0,0 +1,157 @@
+						return f.specialize(functionDefinition, move(_p.first), move(_p.second));
+					}
+				);
+				return move(out) + make_vector<Statement>(move(functionDefinition));
```
```suggestion
				return move(out) + {move(functionDefinition)};
```
Does that work>

hrkrshnn replied:
It doesn't :(

leonardoalt reviewed (commented):
Nice!
